File: gfg/medium/minimize-the-heights3351/solution.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @returns {number}
 */

class Solution {
    getMinDiff(arr, k) {
        arr.sort((a, b) => a - b);

        const n = arr.length;
        let ans = arr[n - 1] - arr[0];

        for (let i = 1; i < n; i++) {
            // Skip if height becomes negative
            if (arr[i] - k < 0) continue;

            const mini = Math.min(arr[0] + k, arr[i] - k);
            const maxi = Math.max(arr[i - 1] + k, arr[n - 1] - k);

            ans = Math.min(ans, maxi - mini);
        }

        return ans;
    }
}

File: gfg/medium/minimize-the-heights3351/solution.py
class Solution:
    def getMinDiff(self, arr, k):
        arr.sort()
        n = len(arr)

        ans = arr[-1] - arr[0]

        for i in range(1, n):
            # Skip if height becomes negative
            if arr[i] - k < 0:
                continue

            mini = min(arr[0] + k, arr[i] - k)
            maxi = max(arr[i - 1] + k, arr[-1] - k)

            ans = min(ans, maxi - mini)

        return ans